File: src/api/parcels.ts
/**
 * Parcels API service for BhoomiSetu.
 * Matches backend schemas from backend/app/schemas_v1.py and backend/app/api/routes/parcels.py.
 */

import { requestJson } from './client'
import type { Parcel as UiParcel } from '../data'

export type ApiParcelResponse = {
  id: string
  parcel_id: string
  project_id: string
  project_code: string | null
  survey_number: string
  state: string
  district: string
  land_area_ha: number
  land_type: string
  land_use: string
  acquisition_status: string
  baseline_risk_score: number
  suitability_score: number
  estimated_delay_months: number
  land_value_inr: number
  estimated_compensation_inr: number
  created_at: string
  updated_at: string
}

export type ApiCurrentAiResult = {
  id: string
  model_version: string
  risk_score: number
  risk_probability: number
  risk_level: string
  acquisition_risk: number
  top_positive_contributors: Array<{ feature: string; impact: number; direction: string }>
  top_negative_contributors: Array<{ feature: string; impact: number; direction: string }>
  is_current?: boolean
  created_at?: string
}

export type ApiParcelDetailResponse = ApiParcelResponse & {
  project?: {
    id: string
    code: string
    name: string
    project_type: string
    district: string
    total_parcels_target: number
    acquisition_progress_pct: number
    status: string
    target_completion_date: string | null
    created_at: string
    parcel_count: number
  } | null
  owners: Array<{
    id: string
    parcel_id: string
    owner_name: string
    share_percentage: number
    is_primary: boolean
    contact_phone: string | null
    created_at: string
  }>
  ml_features: Record<string, unknown>
  current_ai_result: ApiCurrentAiResult | null
  disputes: Array<Record<string, unknown>>
  compensations: Array<Record<string, unknown>>
  documents: Array<Record<string, unknown>>
  geometry_summary?: {
    srid: number
    type: string
    centroid: [number, number] | null
    map_ui_x: number | null
    map_ui_y: number | null
  } | null
}

export type PaginatedParcelsResponse = {
  items: ApiParcelResponse[]
  page: number
  page_size: number
  total: number
}

export type FetchParcelsParams = {
  project?: string
  district?: string
  acquisition_status?: string
  risk_level?: string
  search?: string
  page?: number
  page_size?: number
}

function formatCurrencyCrores(amountInr: number): string {
  if (!amountInr || amountInr === 0) return 'Rs 0'
  const crores = amountInr / 10000000
  return `Rs ${crores.toFixed(2)} Cr`
}

export function mapApiParcelToUi(apiParcel: ApiParcelResponse | ApiParcelDetailResponse): UiParcel {
  const primaryOwner = ('owners' in apiParcel && apiParcel.owners?.find((o) => o.is_primary)?.owner_name)
    || ('owners' in apiParcel && apiParcel.owners?.[0]?.owner_name)
    || 'S. K. Prasad'

  const hasDispute = ('disputes' in apiParcel && (apiParcel.disputes?.length || 0) > 0)
    || apiParcel.baseline_risk_score > 60

  const geom = ('geometry_summary' in apiParcel) ? apiParcel.geometry_summary : null
  const mapX = (geom && typeof geom.map_ui_x === 'number') ? geom.map_ui_x : 50
  const mapY = (geom && typeof geom.map_ui_y === 'number') ? geom.map_ui_y : 50


  const riskScore = Math.round(apiParcel.baseline_risk_score)
  const color = riskScore > 60 ? '#d8634d' : riskScore > 30 ? '#e9a23b' : '#54a884'

  return {
    id: apiParcel.parcel_id,
    survey: apiParcel.survey_number,
    district: apiParcel.district,
    area: `${apiParcel.land_area_ha.toFixed(2)} ha`,
    status: apiParcel.acquisition_status,
    risk: riskScore,
    suitability: Math.round(apiParcel.suitability_score),
    delay: `${apiParcel.estimated_delay_months.toFixed(1)} months`,
    x: mapX,
    y: mapY,
    color: color,
    owner: primaryOwner,
    landType: apiParcel.land_type,
    landUse: apiParcel.land_use,
    project: apiParcel.project_code || 'NH-327 Ring Road',
    value: formatCurrencyCrores(apiParcel.land_value_inr),
    dispute: hasDispute,
  }
}

export async function fetchParcels(params: FetchParcelsParams = {}): Promise<{
  raw: ApiParcelResponse[]
  ui: UiParcel[]
  total: number
  page: number
  page_size: number
}> {
  const queryParts: string[] = []

  if (params.project) queryParts.push(`project=${encodeURIComponent(params.project)}`)
  if (params.district) queryParts.push(`district=${encodeURIComponent(params.district)}`)
  if (params.acquisition_status && params.acquisition_status !== 'All statuses') {
    queryParts.push(`acquisition_status=${encodeURIComponent(params.acquisition_status)}`)
  }
  if (params.risk_level && params.risk_level !== 'All risk levels') {
    queryParts.push(`risk_level=${encodeURIComponent(params.risk_level)}`)
  }
  if (params.search) queryParts.push(`search=${encodeURIComponent(params.search)}`)
  if (params.page) queryParts.push(`page=${params.page}`)
  if (params.page_size) queryParts.push(`page_size=${params.page_size}`)

  const queryString = queryParts.length > 0 ? `?${queryParts.join('&')}` : ''
  const data = await requestJson<PaginatedParcelsResponse>(`/api/v1/parcels${queryString}`)
  
  const raw = data.items || []
  const ui = raw.map(mapApiParcelToUi)
  
  return {
    raw,
    ui,
    total: data.total || raw.length,
    page: data.page || 1,
    page_size: data.page_size || 20,
  }
}

export async function fetchParcelDetail(parcelIdOrCode: string): Promise<{
  raw: ApiParcelDetailResponse
  ui: UiParcel
}> {
  const raw = await requestJson<ApiParcelDetailResponse>(`/api/v1/parcels/${encodeURIComponent(parcelIdOrCode)}`)
  const ui = mapApiParcelToUi(raw)
  return { raw, ui }
}

File: backend/app/api/routes/parcels.py
"""Parcels API routes."""

import uuid
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from geoalchemy2.shape import to_shape
from sqlalchemy import func, or_
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.models.ai_analysis import AIAnalysisResult
from app.models.compensation import CompensationRecord
from app.models.disputes import Dispute
from app.models.documents import DocumentRecord
from app.models.parcels import Parcel, ParcelGeometry, ParcelOwner
from app.models.projects import Project
from app.schemas_v1 import (
    AIAnalysisResponse,
    CompensationResponse,
    DisputeResponse,
    DocumentResponse,
    PaginatedResponse,
    ParcelDetailResponse,
    ParcelOwnerResponse,
    ParcelResponse,
    ProjectResponse,
)

router = APIRouter(prefix="/parcels", tags=["parcels"])




@router.get("", response_model=PaginatedResponse[ParcelResponse])
def get_parcels(
    project: Optional[str] = Query(None, description="Filter by project code or project ID"),
    district: Optional[str] = Query(None, description="Filter by district"),
    acquisition_status: Optional[str] = Query(None, description="Filter by acquisition status"),
    risk_level: Optional[str] = Query(None, description="Filter by AI risk level (if available)"),
    search: Optional[str] = Query(None, description="Search parcel ID or survey number"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
) -> PaginatedResponse[ParcelResponse]:
    """Retrieve paginated land parcels with search and filters."""
    query = db.query(Parcel).join(Project, Parcel.project_id == Project.id)

    if project:
        query = query.filter(
            or_(
                func.lower(Project.code) == project.lower(),
                Parcel.project_id == (uuid.UUID(project) if is_valid_uuid(project) else None),
            )
        )

    if district:
        query = query.filter(func.lower(Parcel.district) == district.lower())

    if acquisition_status:
        query = query.filter(func.lower(Parcel.acquisition_status) == acquisition_status.lower())

    if risk_level:
        # Join with AI analysis result if available
        query = query.outerjoin(AIAnalysisResult, Parcel.id == AIAnalysisResult.parcel_id).filter(
            func.lower(AIAnalysisResult.risk_level) == risk_level.lower()
        )

    if search:
        search_pattern = f"%{search.lower()}%"
        query = query.filter(
            or_(
                func.lower(Parcel.parcel_id).like(search_pattern),
                func.lower(Parcel.survey_number).like(search_pattern),
            )
        )

    total = query.count()
    offset = (page - 1) * page_size
    parcels = query.order_by(Parcel.parcel_id.asc()).offset(offset).limit(page_size).all()

    items = []
    for p in parcels:
        res = ParcelResponse.model_validate(p)
        res.project_code = p.project.code if p.project else None
        items.append(res)

    return PaginatedResponse[ParcelResponse](
        items=items,
        page=page,
        page_size=page_size,
        total=total,
    )


@router.get("/{parcel_id_or_code}", response_model=ParcelDetailResponse)
def get_parcel_detail(
    parcel_id_or_code: str,
    db: Session = Depends(get_db),
) -> ParcelDetailResponse:
    """Retrieve complete parcel profile with project, ownership, 19 ML features, disputes, compensation, and geometry summary."""
    query = db.query(Parcel).filter(
        or_(
            func.lower(Parcel.parcel_id) == parcel_id_or_code.lower(),
            Parcel.id == (uuid.UUID(parcel_id_or_code) if is_valid_uuid(parcel_id_or_code) else None),
        )
    )
    parcel = query.first()

    if not parcel:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Parcel '{parcel_id_or_code}' not found.",
        )

    # 1. Project
    project_res = ProjectResponse.model_validate(parcel.project) if parcel.project else None
    if project_res and parcel.project:
        project_res.parcel_count = db.query(Parcel).filter(Parcel.project_id == parcel.project.id).count()

    # 2. Owners
    owners_res = [ParcelOwnerResponse.model_validate(o) for o in parcel.owners]

    # 3. 19 ML Features exact mapping
    ml_features = {
        "state": parcel.state,
        "district": parcel.district,
        "land_type": parcel.land_type,
        "land_use": parcel.land_use,
        "project_type": parcel.project.project_type if parcel.project else "Road infrastructure",
        "land_area": float(parcel.land_area_ha),
        "land_value": float(parcel.land_value_inr),
        "estimated_compensation": float(parcel.estimated_compensation_inr),
        "number_of_owners": parcel.number_of_owners,
        "ownership_complexity": float(parcel.ownership_complexity_score),
        "previous_dispute": parcel.previous_dispute_flag,
        "previous_objections": parcel.previous_objections_count,
        "environmental_risk": float(parcel.environmental_risk_score),
        "road_accessibility": float(parcel.road_accessibility_score),
        "distance_to_road": float(parcel.distance_to_road_km),
        "stakeholder_count": parcel.stakeholder_count,
        "land_use_conflict": float(parcel.land_use_conflict_score),
        "documentation_completeness": float(parcel.documentation_completeness_score),
        "historical_acquisition_duration": float(parcel.historical_acquisition_duration_months),
    }

    # 4. Current AI Result
    latest_ai = (
        db.query(AIAnalysisResult)
        .filter(AIAnalysisResult.parcel_id == parcel.id, AIAnalysisResult.is_current.is_(True))
        .order_by(AIAnalysisResult.created_at.desc(), AIAnalysisResult.id.desc())
        .first()
    )
    if not latest_ai:
        latest_ai = (
            db.query(AIAnalysisResult)
            .filter(AIAnalysisResult.parcel_id == parcel.id)
            .order_by(AIAnalysisResult.created_at.desc(), AIAnalysisResult.id.desc())
            .first()
        )

    current_ai = None
    if latest_ai:
        current_ai = {
            "id": str(latest_ai.id),
            "model_version": latest_ai.model_version,
            "risk_score": float(latest_ai.risk_score),
            "risk_probability": float(latest_ai.risk_probability),
            "risk_level": latest_ai.risk_level,
            "acquisition_risk": latest_ai.acquisition_risk_class,
            "top_positive_contributors": latest_ai.top_positive_contributors,
            "top_negative_contributors": latest_ai.top_negative_contributors,
            "is_current": latest_ai.is_current,
            "created_at": latest_ai.created_at.isoformat() if latest_ai.created_at else None,
        }


    # 5. Disputes
    disputes = [DisputeResponse.model_validate(d) for d in parcel.disputes]

    # 6. Compensations
    compensations = [CompensationResponse.model_validate(c) for c in parcel.compensations]

    # 7. Documents
    documents = [DocumentResponse.model_validate(doc) for doc in parcel.documents]


    # 8. Geometry Summary
    geom_summary = None
    if parcel.geometry:
        geom_obj = parcel.geometry
        centroid_coords = None
        if geom_obj.centroid:
            try:
                shape_point = to_shape(geom_obj.centroid)
                centroid_coords = [shape_point.x, shape_point.y]
            except Exception:
                centroid_coords = None

        geom_summary = {
            "srid": 4326,
            "type": "Polygon",
            "centroid": centroid_coords,
            "map_ui_x": float(geom_obj.map_ui_x) if geom_obj.map_ui_x is not None else None,
            "map_ui_y": float(geom_obj.map_ui_y) if geom_obj.map_ui_y is not None else None,
        }

    res = ParcelDetailResponse.model_validate(parcel)
    res.project_code = parcel.project.code if parcel.project else None
    res.project = project_res
    res.owners = owners_res
    res.ml_features = ml_features
    res.current_ai_result = current_ai
    res.disputes = disputes
    res.compensations = compensations
    res.documents = documents
    res.geometry_summary = geom_summary

    return res


@router.get("/{parcel_id_or_code}/ai-history", response_model=List[AIAnalysisResponse])
def get_parcel_ai_history(
    parcel_id_or_code: str,
    db: Session = Depends(get_db),
) -> List[AIAnalysisResponse]:
    """Retrieve chronological AI analysis history for a land parcel ordered newest first."""
    query = db.query(Parcel).filter(
        or_(
            func.lower(Parcel.parcel_id) == parcel_id_or_code.lower(),
            Parcel.id == (uuid.UUID(parcel_id_or_code) if is_valid_uuid(parcel_id_or_code) else None),
        )
    )
    parcel = query.first()

    if not parcel:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Parcel '{parcel_id_or_code}' not found.",
        )

    history = (
        db.query(AIAnalysisResult)
        .filter(AIAnalysisResult.parcel_id == parcel.id)
        .order_by(AIAnalysisResult.created_at.desc(), AIAnalysisResult.id.desc())
        .all()
    )

    return [AIAnalysisResponse.model_validate(h) for h in history]


def is_valid_uuid(val: str) -> bool:
    try:
        uuid.UUID(val)
        return True
    except ValueError:
        return False

